iset.mm

```diff
@@ -64015,7 +64015,22 @@ all reals (greatest real number) for which all positive reals are greater.
   $}
 
   ${
-    $d A x $.
+    $d A x $.  $d C x $.
+    $( Left biconditional closure of disjoint union.  (Contributed by Jim
+       Kingdon, 2-Jul-2022.) $)
+    djulclb $p |- ( C e. V -> ( C e. A <-> ( inl ` C ) e. ( A |_| B ) ) ) $=
```

Comment:
May I suggest
```
$p |- ( C e. V -> ( C e. A <-> ( ( inl |` A ) ` C ) e. ( A |_| B ) ) ) $=
```
It's very easy to recover the other version by using ``( F |` _V ) = F``, but for the other direction, you'd have to reprove the result.

Reply:
Part of the problem is that it takes a bit of effort to see how this is used, because the usage is off in #2659 which is waiting for djudom but let me take a look.

So `djuclb` is instantiated as `⊢ (∅ ∈ V → (∅ ∈ 𝑢 ↔ (inl‘∅) ∈ (𝑢 ⊔ 1𝑜)))`. And I don't see how to make that work with ``( ( inl |` u ) ` (/) )`` because the whole point of this part of the proof is that we have `(inl‘∅) ∈ (𝑢 ⊔ 1𝑜)` and are trying to go from there to `∅ ∈ 𝑢`.

This isn't just something I thought up, I'm formalizing the proof of Proposition 1.2 of PradicBrown2021 and that proof is full of `inl(0)` and `inr(0)`.

I probably was too unclear with my earlier comment in favor of the restrictions on `inl` and `inr` - I was really talking about http://us.metamath.org/ileuni/djulf1o.html in particular. More widely, maybe it is good, maybe it is bad, I'm not sure I know enough to say.

Reply:
> I'm formalizing the proof of Proposition 1.2 of PradicBrown2021 and that proof is full of `inl(0)` and `inr(0)`.

It looks to me like the functions used in that proof are (using their notation) `` ( inl |` A ) `` and `` ( inr |` 1o ) ``.

Reply:
> > I'm formalizing the proof of Proposition 1.2 of PradicBrown2021 and that proof is full of `inl(0)` and `inr(0)`.
> 
> It looks to me like the functions used in that proof are (using their notation) `` ( inl |` A ) `` and `` ( inr |` 1o ) ``.

Hmm, I see your point. They do define `inl` (in their notation) as `inl : A → A + B` which seems like there is a different `inl` for each disjoint union in play (which in informal notation can just be glossed over, but for us corresponds to, presumably, restriction).

I'll try formalizing with the restricted `inl` because I guess it probably is feasible (I'll have to try it to convince myself for sure).

Now, even if both are possible, I'm not sure I'm sold on which is better. I want to see what the proofs from #2659 look like with the extra notation. (and even more so if we try to do something difficult, like the proof that http://us.metamath.org/mpeuni/sbth.html implies excluded middle - since all the current work is just the warm-up exercise for that rather longer proof). For the moment I'll keep both `djuclb` and `djuclbN` for the two flavors. But we'll have to figure out whether we keep both (with this naming convention or a better one), consign some to mathboxes, or something else. I think I'm leaning towards some flavor of having both around until we have more experience with this stuff (in the context of set.mm and/or iset.mm - experience from elsewhere is sorta relevant but probably not completely the same), but that's a pretty tentative conclusion.

Reply:
I now believe that ``$p |- ( C e. V -> ( C e. A <-> ( ( inl |` A ) ` C ) e. ( A |_| B ) ) ) $=`` is not provable (or at least, not using any technique we have figured out yet).

The situation is that you have `C e. V` and ``( ( inl |` A ) ` C ) e. ( A |_| B )`` and you need to prove `C e. A`. If `C e. A` were decidable this would be easy, you can show that `-. C e. A` is a contradiction (because a function outside its domain is `(/)` and `(/)` is not an element of any disjoint union). But unless we have some form of reverse closure, I don't see any way to get anything useful out of ``( ( inl |` A ) ` C ) e. ( A |_| B )``.

Reply:
(Obvious preamble: this question is independent from the above question about restriction.)
These questions about "reverse closure" do not appear very natural to me anyway. To me, the mere writing of "``( ( inl |` A )  ` C )``" should be "allowed" only once it has been proved that `C e. A` (possibly, both in a context `ph`).
As a consequence, I wonder if one would need this kind of result for any applications ?

Reply:
Maybe reverse closure isn't the right name for the situation with disjoint union, given that disjoint union carries with it more information, or at least more explicit information, than you have with something like http://us.metamath.org/mpeuni/ndmovrcl.html .

I'm in the process of writing the #2659 proof both ways so we can compare. But basically the situation is either that the proof can carry around (directly, or via a context  and hypotheses) either ``( inl ` C ) e. ( A |_| B )`` or ``A e. C /\ ( ( inl |` A ) ` C ) e. ( A |_| B )``.

The statement that "the mere writing of ``( ( inl |` A ) ` C )`` should be 'allowed' only once it has been proved that C e. A" is perhaps an argument for writing ``( inl ` C )`` instead. But perhaps that has other downsides (which I'll admit I don't really see yet).

Reply:
> The statement that "the mere writing of `` ( ( inl |` A ) ` C ) `` should be 'allowed' only once it has been proved that C e. A" is perhaps an argument for writing `` ( inl ` C ) `` instead.

To me, this is an argument for proving `C e. A` first !  But I admit that I may be a bit of an extremist on that, and a while back, @digama0 remarked that from time to time (especially for the most basic results), we should unapologetically take advantage of some "reverse closure" or related results.

Reply:
> To me, this is an argument for proving `C e. A` first !

A lot of times, I would tend to see this as both appealing and practical. But I'm not so sure when I see how it is interacting with http://us.metamath.org/ileuni/eldju.html in the context of the #2659 proof. Anyway, the #2659 branch now has both versions.